### ISSUE TITLE
New Binary Update

### ISSUE THREAD

**Issue by kkporter** [Simbul/baker#1007]
Hi, 

Any one encounter the following issue during the review process?

I am using bakershelf 4.1 with Magrocket 1.0 and currently this magazine is in production with in-app purchase in Appstore. Now i submitted a new version in itunesconnect, mainly to update my keywords, however on apple review side they reject the new version due to the fact that they are testing to purchase under the sandbox environment, while the backend is already turned into production. therefore the receipt validation failed by the error 21007. 

Any idea how to solve this? We cant switch back the sandbox env cos we have online production users. Does it mean on the backend we need to set up a mirror sandbox environment for review process. In terms of baker and magrocket, what adjustment we need to make? Thanks a lot.

**Comment by freney** [comment]
Wow, that sucks. Have you asked for a review of the rejection, and explain why? I'd be interested to know if Apple realise that it's a problem.

How about a test case in the MagRocket receipt handling, with logic along the lines of this (none of these variables are right, it's just the logic I'm after):

```
if (receiptValidationError == 21007) {
    sandBoxMode = TRUE;
    validateReceipt();
}
```

i.e. if someone tries to do a sandbox purchase, and it fails, switch over to the sandbox server URL and try again.

You can delete that conditional once it's approved.

**Comment by nin9creative** [comment]
I was just going to suggest this......

**Comment by kkporter** [comment]
Thanks a lot, I tried to put this under API index.php,(replace line 736-740) however seems not really working, i guess i am doing it wrongly, please shred some lights, thx

```
if (getiTunesProductionLevel($app_id)=="sandbox") {
    $endpoint = 'https://sandbox.itunes.apple.com/verifyReceipt';
}

else if(getiTunesProductionLevel($app_id)!="sandbox" && $data->status == 21007){
```

$endpoint = 'https://sandbox.itunes.apple.com/verifyReceipt';}
else{
$endpoint = 'https://buy.itunes.apple.com/verifyReceipt';
}

}
    }

**Comment by nin9creative** [comment]
Are you using the MagRocket code from main in Github, or the MagRocket-Backend 1.0 and MagRocket-API 1.0.1?

Andrew

**Comment by kkporter** [comment]
Hi Andrew,

have been using the the MagRocket-Backend 1.0 and MagRocket-API 1.0.1, not the main from github.

**Comment by nin9creative** [comment]
I'll try to look at this tonight.

Andrew

**Comment by kkporter** [comment]
thanks Andrew, take your time!

**Comment by nin9creative** [comment]
Ok, I took a look at this for a few hours today... UGH!!!   LOL  It would have taken 15 minutes if I realized that the getCode in PHP errors returned a string not an integer :)

You can see the modified code for the API index.php here.  I would just look at the following lines in this PASTEBIN file and make the changes in your local copy.  There are changes basically around the calls the verifyReceipt that fall back to using the Sandbox when the specific error of 21007 is encountered (which means that a Sandbox receipt was posted against the Production validation service).

http://pastebin.com/utAvH2sW

*line numbers relative to the PASTEBIN version

/purchases
Lines 308-317

/confirmpurchase
Lines 424-433

verifyReceipt
*basically the whole function as I added an optional parameter for override to Sandbox
Lines 688-743

**Comment by kkporter** [comment]
Thanks a lot, Andrew!

I tried the code and the issue could be downloaded in sandbox environment without any issues, bravo! However, in the sandbox environment, the cover image wouldn't update itself in the iPhone/iPad newsstand shelf as the new issue is downloaded. Not sure if the reviewer would care about that tho :) Is it a quick fix to make the cover image update as well in case of the 21007 condition?

I will try to resubmit for itunesconnect and I will report back to help in case anyone would run into a similar issue.

**Comment by nin9creative** [comment]
Hmmm...  Well, that I don't think is a MagRocket issue.  The ability to purchase it and have it show up as "purchased" is the only bit that the calls to MagRocket
would affect in this case.  Not sure why the cover image wouldn't update...

**Comment by kkporter** [comment]
ok thx, it weird though..let me try to submit and see how it goes,
thanks again!

**Comment by kkporter** [comment]
Hi all,

just to report back that my update binary has been passed with Apple review. So if anyone in the future run into a similar issue when update your binary, and reviewer said they cannot buy the issue under sandbox environment. Please apply the above fix posted by Andrew on the Magrocket backend, and you will be good to go.

Thanks Andrew again, and considered this case closed.

**Comment by folletto** [comment]
Thanks for taking your time to report back, that's a very valuable feedback. :)

**Comment by nin9creative** [comment]
With the upcoming 1.1 Version of MagRocket - this fix will be included (fallback to Sandbox when Production fails with 21007 error).

If anyone needs to fix it prior to that, per @kkporter you can apply the fix.

Andrew
